### DISCUSSION TITLE
Option to reduce 'bloat' to reduce usage for higher thread number

### DISCUSSION THREAD

**Discussion by AlexPaiva** [seleniumbase/SeleniumBase#2393] [Ideas]
A simple option that when enabled removes a lot of bloat from chrome where the goal is to reach a higher thread number where every less mb of consumption counts. Just getting it out there if it's doable, have seen a few people asking for reducing memory and cpu usage on high threads.

**Comment by mdmintz** [comment]
Are you referring to pytest multithreading internals or Chrome internals? `pytest` is a separate repo, and Chrome can be configured via various Selenium options available. Not everything about Chrome can be configured though.

**Reply by AlexPaiva** [reply]
Both, and my question was aswell where can I find the options for that?

**Reply by mdmintz** [reply]
`pytest` has its own repo: https://github.com/pytest-dev/pytest (lots of command-line options).
It also has several plugins, such as `pytest-xdist` (which provides the multithreading options):
https://github.com/pytest-dev/pytest-xdist

For a list of Chromium command-line options, see: https://peter.sh/experiments/chromium-command-line-switches/
SeleniumBase already uses many of them, and has its own args for enabling the ones that aren't enabled by default.
Use SeleniumBase's `chromium_arg` to pass a list of comma-separated Chromium args if there's no direct arg for it.

**Comment by AlexPaiva** [comment]
> `pytest` has its own repo: https://github.com/pytest-dev/pytest (lots of command-line options). It also has several plugins, such as `pytest-xdist` (which provides the multithreading options): https://github.com/pytest-dev/pytest-xdist
> 
> For a list of Chromium command-line options, see: https://peter.sh/experiments/chromium-command-line-switches/ SeleniumBase already uses many of them, and has its own args for enabling the ones that aren't enabled by default. Use SeleniumBase's `chromium_arg` to pass a list of comma-separated Chromium args if there's no direct arg for it.

Where can I find the list of the sb list of args?

**Reply by mdmintz** [reply]
The SeleniumBase pytest args are listed here: [SeleniumBase/help_docs/customizing_test_runs.md](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/customizing_test_runs.md)

**Reply by AlexPaiva** [reply]
I meant for the raw driver

**Reply by mdmintz** [reply]
https://github.com/seleniumbase/SeleniumBase/blob/36d410bd8ccfc47a2332141eabc105cbadedb255/seleniumbase/plugins/driver_manager.py#L65

**Comment by AlexPaiva** [comment]
> https://github.com/seleniumbase/SeleniumBase/blob/36d410bd8ccfc47a2332141eabc105cbadedb255/seleniumbase/plugins/driver_manager.py#L65

Got it, thanks! When I am using a pytest xdist do I select multi_proxy on the raw driver?

**Reply by mdmintz** [reply]
If you are using more than one proxy, then yes, set `multi_proxy=True`.

**Reply by mdmintz** [reply]
More info on using multiple proxies: https://github.com/seleniumbase/SeleniumBase/issues/1832#issue-1661403259

**Reply by AlexPaiva** [reply]
Got it, thanks!

Quick question, for raw driver how do I change window height/width? I assume 'd_width=None,  # Set device width
    d_height=None,  # Set device height
    d_p_r=None,  # Set device pixel ratio' are different and relating to different things? if so, what exactly?

**Reply by mdmintz** [reply]
Those are for mobile device metrics. (Something different, only for `mobile` mode.)

You want this: `driver.set_window_size(width, height)`

(It is set after the browser is launched.)

**Reply by AlexPaiva** [reply]
Ah ok I see, where can I find the list of the '.sets' available for raw driver such as that one?

**Reply by mdmintz** [reply]
If it's already included with raw `selenium` (eg. `driver.set_window_size(width, height)`), then it won't be listed in the SeleniumBase Docs.  (It's probably already somewhere in the Selenium Docs.)

But there is a section for `driver` methods that were added in SeleniumBase: https://github.com/seleniumbase/SeleniumBase/issues/2200#issue-1952880367

**Comment by AlexPaiva** [comment]
In regards to the thread I meant if there was a way to actually remove things that weight the browser down or change the binary itself or something related but I assume that's creating a new browser?

**Reply by mdmintz** [reply]
I'm not sure what you mean by this. Chrome is what it is. There are options available so that you can customize Chrome. SeleniumBase provides command-line options and method args for making those customizations, as listed in the comments above.

**Reply by AlexPaiva** [reply]
I meant modify the Chromium source code itself then rebuild to be used with selenium base or is this just making it detectable and maybe a monkeypatch is better or?

**Reply by mdmintz** [reply]
> "modify the Chromium source code"

I don't know anything about that. UC Mode modifies chromedriver (not Chrome / Chromium).
Most Chromium browsers are just modified versions of Chrome anyway (eg. Brave or Opera).

**Comment by AlexPaiva** [comment]
> > "modify the Chromium source code"
> 
> I don't know anything about that. UC Mode modifies chromedriver (not Chrome / Chromium). Most Chromium browsers are just modified versions of Chrome anyway (eg. Brave or Opera).

Yes that's what I meant, so one could technically modify the chromedriver directly and perform direct changes on it?

**Reply by mdmintz** [reply]
That's what [undetected/patcher.py](https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/undetected/patcher.py) does.

**Reply by AlexPaiva** [reply]
Noted, just a quick question, how do you know here where to and what to patch if you have the time to do a quick walkthrough?

```
def patch_exe(self):
        """Patches the ChromeDriver binary"""
        def gen_js_whitespaces(match):
            return b"\n" * len(match.group())

        def gen_call_function_js_cache_name(match):
            rep_len = len(match.group()) - 3
            ran_len = random.randint(6, rep_len)
            bb = b"'" + bytes(str().join(random.choices(
                population=string.ascii_letters, k=ran_len
            )), 'ascii') + b"';" + (b"\n" * (rep_len - ran_len))
            return bb

        with io.open(self.executable_path, "r+b") as fh:
            file_bin = fh.read()
            file_bin = re.sub(
                b"window\\.cdc_[a-zA-Z0-9]{22}_"
                b"(Array|Promise|Symbol|Object|Proxy|JSON)"
                b" = window\\.(Array|Promise|Symbol|Object|Proxy|JSON);",
                gen_js_whitespaces,
                file_bin,
            )
            file_bin = re.sub(
                b"window\\.cdc_[a-zA-Z0-9]{22}_"
                b"(Array|Promise|Symbol|Object|Proxy|JSON) \\|\\|",
                gen_js_whitespaces,
                file_bin,
            )
            file_bin = re.sub(
                b"'\\$cdc_[a-zA-Z0-9]{22}_';",
                gen_call_function_js_cache_name,
                file_bin,
            )
            fh.seek(0)
            fh.write(file_bin)
        return True
```

**Reply by mdmintz** [reply]
You should ask the maintainer of `undetected-chromedriver`.
It came from [undetected_chromedriver/patcher.py](https://github.com/ultrafunkamsterdam/undetected-chromedriver/blob/master/undetected_chromedriver/patcher.py)
